Replace AdalHttpClient with HttpManager

bgavrilMS opened the pull request.

jennyf19 committed:
initial commit

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/AuthenticationContext.cs:
```diff
@@ -533,7 +536,7 @@ internal async Task<AuthenticationResult> AcquireTokenForClientCommonAsync(strin
             };
 
-            var handler = new AcquireTokenSilentHandler(requestData, userId, parameters);
+            var handler = new AcquireTokenSilentHandler(_serviceBundle, requestData, userId, parameters);
```
Look ok, service bundle should be made available to all objects.  #Closed

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/AuthenticationParameters.cs:
```diff
@@ -60,13 +61,23 @@ public sealed class AuthenticationParameters
 
+        private readonly IHttpManager _httpManager;
+
+        internal AuthenticationParameters()
```
This ctor implies that _httpManager can be null. Can it? #Resolved

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/OAuth2/OAuthClient.cs:
```diff
@@ -210,5 +194,111 @@ private static string CheckForExtraQueryParameter(string url)
             return url;
         }
+
+        public async Task<IHttpWebResponse> ExecuteRequestAsync<T>()
```
The difference between this and GetResponseAsync is not clear.  #Resolved

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/OAuth2/OAuthClient.cs:
```diff
@@ -210,5 +194,111 @@ private static string CheckForExtraQueryParameter(string url)
+            return webResponse;
+        }
+
+        public static IHttpWebResponse CreateResponse(HttpResponse response)
```
Please try to lower visibility of methods. Too many public methods - are they really needed? #Resolved

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireDeviceCodeHandler.cs:
```diff
@@ -90,13 +93,11 @@ private DictionaryRequestParameters GetRequestParameters()
         {
             await _authenticator.UpdateFromTemplateAsync(_requestContext).ConfigureAwait(false);
-            AdalHttpClient client = new AdalHttpClient(_authenticator.DeviceCodeUri, _requestContext)
+            OAuthClient client = new OAuthClient(_serviceBundle.HttpManager, _authenticator.DeviceCodeUri, _requestContext)
```
Nice #Closed

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/Authenticator.cs:
```diff
@@ -98,9 +99,12 @@ public string GetAuthorityHost()
 
         public async Task UpdateFromTemplateAsync(RequestContext requestContext)
         {
+            InstanceDiscovery instanceDiscovery = new InstanceDiscovery(ServiceBundle.HttpManager);
```
There should be one already in ServiceBundle. #Resolved

bgavrilMS commented:
Looks good so far, I imagine tests need work.  #Closed

jennyf19 committed:
fix merge issues

henrik-me commented on devApps/AdalDesktopTestApp/MyHttpClientFactory.cs:
```diff
@@ -0,0 +1,31 @@
+
+        public HttpClient GetHttpClient()
+        {
+            _httpClient = new HttpClient(new HttpClientHandler() { UseDefaultCredentials = true })
```
Suggestion: return the same httpclient for the lifetime of the factory? If you need something that creates a new object for testing purposes, you can add a factory method which creates a new one (CreateHttpClient) #Resolved

bgavrilMS replied:
You can use Lazy<T> if you want #Resolved

henrik-me commented on devApps/AdalDesktopTestApp/Program.cs:
```diff
@@ -142,15 +140,16 @@ private static async Task RunAppAsync(AuthenticationContext context)
         {
-            HttpClient httpClient = CreateCustomHttpClient();
+            IHttpClientFactory myHttpClientFactory = new MyHttpClientFactory();
+            myHttpClientFactory.GetHttpClient();
```
>myHttpClientFactory.GetHttpClient(); [](start = 12, length = 36)

what effect does this really have? #Resolved

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/IServiceBundle.cs:
```diff
@@ -35,5 +37,8 @@ internal interface IServiceBundle
         IPlatformProxy PlatformProxy { get; }
         IWsTrustWebRequestManager WsTrustWebRequestManager { get; }
+
+        InstanceDiscovery InstanceDiscovery { get; }
```
nit: what does this empty line do? #Resolved

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Http/RedirectUriHelper.cs:
```diff
@@ -55,7 +55,7 @@ public static void Validate(Uri redirectUri, bool usesSystemBrowser = false)
             }
 
-            // Currenlty only MSAL supports the system browser, on Android and iOS
+            // Currently only MSAL supports the system browser, on Android and iOS
```
>MSAL [](start = 30, length = 4)

not this PR: is this code then still relevant as it's not in ADAL? Is this dead code? #Resolved

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/OAuth2/OAuthClient.cs:
```diff
@@ -0,0 +1,255 @@
+
+namespace Microsoft.Identity.Core.OAuth2
+{
+    internal class OAuthClient
```
>OAuthClient [](start = 19, length = 11)

is this a copy of what is in MSAL? is it a copy of the AdalHttpClient? or does it take extra responsibilities?

jennyf19 replied:
this is closer to a copy of AdalHttpClient, but with a name similar to what we have in MSAL (but can change the name). 

---
In reply to: [267507862](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1530#discussion_r267507862) [](ancestors = 267507862)

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Exceptions/AdalServiceException.cs:
```diff
@@ -74,37 +74,35 @@ internal AdalServiceException(string errorCode, Exception innerException)
         /// <param name="innerException">The exception that is the cause of the current exception, or a null reference if no inner exception is specified. It may especially contain the actual error message returned by the service.</param>
-        internal AdalServiceException(string errorCode, string message, string[] serviceErrorCodes, Exception innerException)
-            : base(errorCode, message, (innerException is HttpRequestWrapperException) ? innerException.InnerException : innerException)
+        internal AdalServiceException(string errorCode, string message, string[] serviceErrorCodes, IHttpWebResponse response, Exception innerException)
```
>string[] serviceErrorCodes, IHttpWebResponse response, [](start = 72, length = 54)

note: it made me a bit confused reading the code that innerException is moving location, though this also makes sense. What is the consideration for keeping innerException at the end?

jennyf19 replied:
just for consistency, as it was at the end before, but now we need the IHttpWebResponse. i could move that to the end though. Do you have a preference?

---
In reply to: [267508974](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1530#discussion_r267508974) [](ancestors = 267508974)

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs:
```diff
@@ -220,12 +223,27 @@ public async Task<AuthenticationResult> RunAsync()
+                    if (serviceException.StatusCode >= 500 && serviceException.StatusCode < 600 ||
+                    serviceException.StatusCode == 408)
+                    {
+                        _client.Resiliency = true;
```
>_client.Resiliency = true; [](start = 24, length = 26)

consider changing this to a simple flag, then you don't have to have the create OAuthClient above #Resolved

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs:
```diff
@@ -220,12 +223,27 @@ public async Task<AuthenticationResult> RunAsync()
+                if (_client == null)
+                {
+                    _client = new OAuthClient();
+                }
```
see below, you have an opportunity to replace this with a flag instead of newing up this object #Resolved

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs:
```diff
@@ -220,12 +223,27 @@ public async Task<AuthenticationResult> RunAsync()
-                    RequestContext.Logger.Info("Refreshing access token failed due to one of these reasons:- Internal Server Error, Gateway Timeout and Service Unavailable. " +
-                                       "Hence returning back stale access token");
+                    if (serviceException.StatusCode >= 500 && serviceException.StatusCode < 600 ||
+                    serviceException.StatusCode == 408)
```
>408 [](start = 51, length = 3)

consider making this refer to a specific httpSTatus  code from the enum (timeout). I believe the enum already exists HttpStatusCode #Resolved

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs:
```diff
@@ -301,20 +319,20 @@ protected virtual async Task PreRunAsync()
+            if (!Authenticator.Authority.Equals(updatedAuthority, StringComparison.OrdinalIgnoreCase))
             {
-                await Authenticator.UpdateAuthorityAsync(updatedAuthority, RequestContext).ConfigureAwait(false);
+                await Authenticator.UpdateAuthorityAsync(updatedAuthority, RequestContext, ServiceBundle).ConfigureAwait(false);
```
>ServiceBundle [](start = 91, length = 13)

nit: in other places we have service bundle first #Resolved

jennyf19 replied:
yep. thanks. we have it first in MSAL, so was going for consistency. missed this one.

---
In reply to: [267510352](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1530#discussion_r267510352) [](ancestors = 267510352)

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenOnBehalfHandler.cs:
```diff
@@ -68,7 +71,7 @@ public AcquireTokenOnBehalfHandler(RequestData requestData, UserAssertion userAs
         {
             requestParameters[OAuthParameter.GrantType] = OAuthGrantType.JwtBearer;
-            requestParameters[OAuthParameter.Assertion] = this.userAssertion.Assertion;
+            requestParameters[OAuthParameter.Assertion] = this._userAssertion.Assertion;
```
>this. [](start = 58, length = 5)

remove this. #Resolved

henrik-me reviewed (changes requested):
:clock1:

henrik-me commented:
still reviewing

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/AuthenticationParameters.cs:
```diff
@@ -60,13 +61,20 @@ public sealed class AuthenticationParameters
         /// </summary>
         public string Resource { get; set; }
 
+        private static IHttpManager HttpManager { get; set; }
```
You don't need a settter if you only set it from the ctor. In fact, it would be better if this was a readonly member. #Resolved

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/AuthenticationParameters.cs:
```diff
@@ -147,17 +155,17 @@ private static async Task<AuthenticationParameters> CreateFromResourceUrlCommonA
-                IHttpClient request = new HttpClientWrapper(resourceUrl.AbsoluteUri, null);
-                await request.GetResponseAsync().ConfigureAwait(false);
+                OAuthClient client = new OAuthClient(HttpManager, resourceUrl.AbsoluteUri, null);
+                await client.ExecuteRequestAsync<IHttpWebResponse>().ConfigureAwait(false);
```
Why do I need to pass a type as a generic param ? i expect that ExecuteRequestAsync to always return IHttpWebResponse.  #Resolved

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Exceptions/ExceptionDetail.cs:
```diff
@@ -62,7 +64,8 @@ public static ExceptionDetail FromHttpResponse(IHttpWebResponse response)
                 ResponseBody = response?.Body,
                 StatusCode = response != null ? (int)response.StatusCode : -1,
-                HttpResponseHeaders = response?.Headers
+                HttpResponseHeaders = response?.Headers,
```
nit: if we record the response here, we don't need to separately record the headers. #Resolved

jennyf19 replied:
ignoring for now.

---
In reply to: [267526403](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1530#discussion_r267526403) [](ancestors = 267526403)

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/OAuth2/OAuthClient.cs:
```diff
@@ -0,0 +1,255 @@
+        private readonly IHttpManager _httpManager;
+        private readonly Dictionary<string, string> _headers = new Dictionary<string, string>();
+
+        internal OAuthClient() { }
```
Is this ctor valid? Can this class function without an http manager? #Resolved

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/OAuth2/OAuthClient.cs:
```diff
@@ -0,0 +1,255 @@
+
+        public int TimeoutInMilliSeconds { set; get; } = 30000;
+        public IRequestParameters BodyParameters { get; set; }
+        internal string RequestUri { get; set; }
```
Please review all the setters - if you set it from the constructor, you don't need to add "set". If you set it only from this class, do interal string Prop { get; private set; }. This greatly helps understanding who is allowed to set this property. Ideally all properties and members are readonly, i.e. only the ctor is allowed to set them. #Resolved

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/OAuth2/OAuthClient.cs:
```diff
@@ -0,0 +1,255 @@
+            return url;
+        }
+
+        internal async Task<IHttpWebResponse> ExecuteRequestAsync<T>()
```
No T needed. #Resolved

bgavrilMS replied:
So this class exposes 2 public methods that do almost the same thing - one that executes a request and serializes to an object T, and one that just executes the request. I am wondering if the second one is needed. #Resolved

bgavrilMS commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs:
```diff
@@ -220,12 +223,27 @@ public async Task<AuthenticationResult> RunAsync()
                 RequestContext.Logger.ErrorPii(ex);
-                if (_client != null && _client.Resiliency && extendedLifetimeResultEx != null)
+
+                if (_client == null)
```
Why not just creating a client in the ctor? #Resolved

bgavrilMS commented on tests/Test.ADAL.NET.Unit.net45/BrokerParametersTests.cs:
```diff
@@ -60,13 +56,15 @@ public class BrokerParametersTests
             ExtendedLifeTimeEnabled = false
         };
 
+        IServiceBundle _serviceBundle = TestCommon.CreateDefaultServiceBundle();
```
needed?
 #Resolved

jennyf19 committed:
first round of PR comments:
- remove system browser check in Validate
- clean up setters and stuff
- clean up ATHandlerBase and resiliency
- put serviceBunny first

jennyf19 commented:
i know. :)

---
In reply to: [474997042](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1530#issuecomment-474997042) [](ancestors = 474997042)

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenOnBehalfHandler.cs:
```diff
@@ -68,10 +71,8 @@ public AcquireTokenOnBehalfHandler(RequestData requestData, UserAssertion userAs
         {
             requestParameters[OAuthParameter.GrantType] = OAuthGrantType.JwtBearer;
-            requestParameters[OAuthParameter.Assertion] = this.userAssertion.Assertion;
+            requestParameters[OAuthParameter.Assertion] = this._userAssertion.Assertion;
```
>this [](start = 58, length = 4)

this. is still around #Resolved

jennyf19 committed:
pr comments

jennyf19 committed:
move servicebundle in test

jennyf19 commented:
UI automation build passed [here](https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=312441)

jennyf19 reviewed (approved):
:shipit:  from @bgavrilMS 😆

henrik-me reviewed (approved):
:shipit:
